### PR TITLE
feat: Add property: private? for attributes, relationships, aggregates, and calculations

### PR DESCRIPTION
**Pull request by kyle5794**
# Contributor checklist

- [ ] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

### CODE REVIEW

**Comment by kyle5794** [comment]
Reference ash-project/ash_json_api#33

File: lib/ash/resource/attribute.ex [hunk 1]
```diff
@@ -62,6 +64,11 @@ defmodule Ash.Resource.Attribute do
       default: true,
       doc: "Whether or not the value can be written to"
     ],
+    private?: [
+      type: :boolean,
+      default: false,
+      doc: "Whether or not the value can be accessed by extensions"
```
Comment (zachdaniel) [inline lib/ash/resource/attribute.ex:70]: I might say "Whether or not the attribute will appear in any interfaces created off of this resource, e.g `AshJsonApi` and `AshGraphql`"

File: lib/ash/resource/transformers/create_join_relationship.ex [hunk 2]
```diff
@@ -26,7 +26,8 @@ defmodule Ash.Resource.Transformers.CreateJoinRelationship do
           name: relationship.join_relationship,
           destination: relationship.through,
           destination_field: relationship.source_field_on_join_table,
-          source_field: relationship.source_field
+          source_field: relationship.source_field,
+          private?: relationship.private?
```
Comment (zachdaniel) [inline lib/ash/resource/transformers/create_join_relationship.ex:30]: This should actually just be `private?: true`. If they want the join relationship to be public they can declare the join relationship themselves and point the `many_to_many` at that join relationship that they created.

File: lib/ash/resource/attribute.ex (hunk 1 above)
Comment (kyle5794) [reply lib/ash/resource/attribute.ex]: Updated.

**Comment by zachdaniel** [comment]
@kyle5794 one more thing here: Lets add `Ash.Resource.public_attribute/2`, `Ash.Resource.public_attributes/1`, `Ash.Resource.public_relationship/2` and `Ash.Resource.public_relationships/1`. Then when we use this in the extensions, we can make sure they only use that, as opposed to their standard counterparts.

File: lib/ash/resource/transformers/belongs_to_attribute.ex [hunk 3]
```diff
@@ -24,6 +24,7 @@ defmodule Ash.Resource.Transformers.BelongsToAttribute do
              name: relationship.source_field,
              type: relationship.field_type,
              writable?: false,
+             private?: relationship.private?,
```
Comment (zachdaniel) [inline lib/ash/resource/transformers/belongs_to_attribute.ex:27]: This should also just be `private?: true`.

**Comment by zachdaniel** [comment]
@kyle5794 alright last round of feedback, this time for real :)

**Comment by kyle5794** [comment]
@zachdaniel I wonder if we need to add `private?` for `Ash.Resource.Aggregate` too?

**Comment by zachdaniel** [comment]
Hmm...yeah, probably. And calculations too.

**Comment by zachdaniel** [comment]
Great work! Next we'll want to get this in use by AshJsonApi and AshGraphql :) 

🚀 Thank you for your contribution! 🚀

**Comment by zachdaniel** [comment]
@kyle5794 I'm going to add this to ash_graphql if you want to add it to ash_json_api :)

**Comment by kyle5794** [comment]
@zachdaniel that works for me too. However, I can't start working on that yet until this saturday. Hope that'll be ok :)

**Comment by zachdaniel** [comment]
Sounds good! If I do it in the meantime I'll let you know.

**Comment by zachdaniel** [comment]
@kyle5794 i only got around to ash graphql so ash json api is still open :)

**Comment by kyle5794** [comment]
@zachdaniel got it. I'll work on it. :+1:

**Pull request merged**